src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -305,66 +263,79 @@ end = struct
-             , pending_coinbases ) =
-      Mvar.take result
-    in
-    assert (
-      Ledger.Db.(
-        Ledger_hash.equal (merkle_root ledger_db) (merkle_root synced_db)
-        && Ledger_hash.equal (merkle_root ledger_db)
-             (Frozen_ledger_hash.to_ledger_hash snarked_ledger_hash)) ) ;
-    let new_root =
-      With_hash.map t.current_root ~f:(fun root ->
-          (* Need to coerce new_root from a proof_verified transition to a
-             fully verified transition because it will be added into transition
-             frontier*)
-          let (`I_swear_this_is_safe_see_my_comment verified_root) =
-            External_transition.(root |> of_proof_verified |> to_verified)
-          in
-          verified_root )
+      (synced_db, data)
     in
     match%bind
-      Staged_ledger.of_scan_state_pending_coinbases_and_snarked_ledger
-        ~scan_state
-        ~snarked_ledger:(Ledger.of_database synced_db)
-        ~expected_merkle_root ~pending_coinbases
+      Network.get_staged_ledger_aux_and_pending_coinbases_at_hash t.network
+        sender hash
     with
-    | Ok root_staged_ledger ->
-        let%map new_frontier =
-          Transition_frontier.create ~logger ~root_transition:new_root
-            ~root_snarked_ledger:synced_db ~root_staged_ledger
-            ~consensus_local_state:
-              (Transition_frontier.consensus_local_state frontier)
+    | Error _ -> failwith "..."
+    | Ok (scan_state, expected_merkle_root, pending_coinbases) ->
+        assert (
+          Ledger.Db.(
+            Ledger_hash.equal (merkle_root ledger_db) (merkle_root synced_db)
+            (*&& Ledger_hash.equal (merkle_root ledger_db)
+             (Frozen_ledger_hash.to_ledger_hash snarked_ledger_hash) *))
+        ) ;
+        let new_root =
+          With_hash.map t.current_root ~f:(fun root ->
+              (* Need to coerce new_root from a proof_verified transition to a
+             fully verified transition because it will be added into transition
+             frontier*)
+              let (`I_swear_this_is_safe_see_my_comment verified_root) =
+                External_transition.(root |> of_proof_verified |> to_verified)
+              in
+              verified_root )
+        in
+        let received_staged_ledger_hash =
+          Staged_ledger_hash.of_aux_ledger_and_coinbase_hash
+            (Staged_ledger_hash.Aux_hash.of_bytes
+               (Staged_ledger_aux_hash.to_bytes
+                  (Staged_ledger.Scan_state.hash scan_state)))
+            expected_merkle_root pending_coinbases
         in
-        Logger.info logger ~module_:__MODULE__ ~location:__LOC__
-          "Bootstrap state: complete." ;
-        (new_frontier, Transition_cache.data transition_graph)
-    | Error err ->
-        (* TODO: punish *)
-        Logger.faulty_peer logger ~module_:__MODULE__ ~location:__LOC__
-          "received faulty scan state from the peer." ;
-        Error.raise err
+        if
```

Comment:
Rather than having this if statement, convert the boolean expression into a Result.t via `ok_if_true`. Using this, you can Result.bind to make the code cleaner.